src/consensus/encode.rs

```diff
@@ -448,8 +449,18 @@ impl<T: Encodable> Encodable for Vec<T> {
 impl<T: Decodable> Decodable for Vec<T> {
     #[inline]
     fn consensus_decode<D: io::Read>(d: &mut D) -> Result<Self, Error> {
-        let len = VarInt::consensus_decode(d)?.0;
-        let mut ret = Vec::with_capacity(len as usize);
+        let len = usize::try_from(*VarInt::consensus_decode(d)?)
+            .map_err(|_| self::Error::ParseFailed("VarInt overflows usize"))?;
+
+        // Prevent a too large allocation, see docs for `Vec::with_capacity
+        let layout_size = mem::size_of::<T>().saturating_mul(len);
+        if layout_size > isize::MAX as usize {
+            return Err(self::Error::ParseFailed(
+                "length exceeds maximum allocatable bytes",
+            ));
+        }
```

Comment:
`isize::MAX` is still an absurd amount of memory.

`rust-bitcoin` - which is the origin of most of this decoding code - had to deal with the same problem and solved it by capping the max allocation to 4MB: https://github.com/rust-bitcoin/rust-bitcoin/blob/e99177c4fba9ece1eb219e7980981c099bc6f2d5/src/consensus/encode.rs#L307-L308

I think we should do something similar and cap the maximum allocation at a much smaller number than `isize::MAX`.

Reply:
4MB looks reasonable, I asked on `#monero-dev` if such limit exists in the codebase and it looks like there isn't so I'm fine with any 'arbitrary' number that make sense!

Reply:
Agree with this - I've been AFK for the last week but I'll push this change now